tests/switchless_one_tcs/host/host.c

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <inttypes.h>
+#include <limits.h>
+#include <openenclave/host.h>
+#include <openenclave/internal/atomic.h>
+#include <openenclave/internal/error.h>
+#include <openenclave/internal/tests.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include "../../../host/hostthread.h"
+#include "../../../host/strings.h"
+#include "switchless_one_tcs_u.h"
+
+#if defined(__linux__)
+#include <unistd.h>
+#elif defined(_WIN32)
+#include <windows.h>
+#endif
+
+static status_e switchless_func_status = FUNC_NOT_START;
+void* thread_func(void* arg)
+{
+    oe_enclave_t* enclave = (oe_enclave_t*)arg;
+    oe_result_t ret = enc_empty_switchless(enclave, &switchless_func_status);
+    OE_TEST(ret == OE_OK);
+    return NULL;
+}
+
+int main(int argc, const char* argv[])
+{
+    oe_enclave_t* enclave = NULL;
+    oe_result_t result;
+
+    if (argc != 2)
+    {
+        fprintf(stderr, "Usage: %s ENCLAVE\n", argv[0]);
+        exit(1);
+    }
+
+    printf("OneTCSTest::test\n");
+    printf("Create enclave with Switchless Calls and one TCS only.\n");
+    printf("Call empty switchless ecall, it is expected to pass...\n");
+    printf("Call empty ecall, it is expected to fail since TCS is occupied by "
+           "trusted worker thread...\n");
+
+    const uint32_t flags = oe_get_create_flags();
+
+    oe_enclave_setting_context_switchless_t switchless_setting = {1, 1};
+    oe_enclave_setting_t settings[] = {
+        {.setting_type = OE_ENCLAVE_SETTING_CONTEXT_SWITCHLESS,
+         .u.context_switchless_setting = &switchless_setting}};
+
+    if ((result = oe_create_switchless_one_tcs_enclave(
+             argv[1],
+             OE_ENCLAVE_TYPE_SGX,
+             flags,
+             settings,
+             OE_COUNTOF(settings),
+             &enclave)) != OE_OK)
+        oe_put_err("oe_create_enclave(): result=%u", result);
+
+    oe_thread_t tid;
+    int ret = 0;
+    if ((ret = oe_thread_create(&tid, thread_func, enclave)))
+    {
+        oe_put_err("thread_create(host): ret=%u", ret);
+    }
+
+    while (switchless_func_status != FUNC_WORKING)
+    {
+#if defined(__linux__)
+        usleep(100);
+#elif defined(_WIN32)
+        Sleep(0);
+#endif
+    }
```

Comment:
Can we use a condition variable instead of busy waiting?

Reply:
Also, prefer C++ functions to OS-specific C functions on the host to avoid ifdef's

Reply:
I found that oe_cond_wait calls ocall , maybe TCS is not occupied. so OE_TEST(enc_empty_regular(enclave) == OE_OUT_OF_THREADS) will fail.

static int _thread_wait(oe_thread_data_t* self)
{
    const void* tcs = td_to_tcs((oe_sgx_td_t*)self);

    if (oe_ocall(OE_OCALL_THREAD_WAIT, (uint64_t)tcs, NULL) != OE_OK)
        return -1;

    return 0;
}

static int _thread_wake_wait(oe_thread_data_t* waiter, oe_thread_data_t* self)
{
    int ret = -1;
    uint64_t waiter_tcs = (uint64_t)td_to_tcs((oe_sgx_td_t*)waiter);
    uint64_t self_tcs = (uint64_t)td_to_tcs((oe_sgx_td_t*)self);

    if (oe_sgx_thread_wake_wait_ocall(oe_get_enclave(), waiter_tcs, self_tcs) !=
        OE_OK)
        goto done;

    ret = 0;

done:
    return ret;
}

Reply:
I don't understand, wouldn't the waiting be happening on the host, not the enclave?

Reply:
You mentioned the busy wait, I understand it  happens on the enclave, is it right?

> void enc_empty_switchless(status_e *status)
{
    int num_waiting = 0;
    *status = FUNC_WORKING;
    while (*status != FUNC_EXIT)
    {
        num_waiting++;
    }
    oe_host_printf("enc_empty_switchless exit\n");
}

Reply:
Sorry, I meant this code:

```
    while (switchless_func_status != FUNC_WORKING)
    {
#if defined(__linux__)
        usleep(100);
#elif defined(_WIN32)
        Sleep(0);
#endif
    }
```